src/wallet/salvage.cpp

```diff
@@ -0,0 +1,150 @@
+    //  ... repeated
+    // DATA=END
+
+    std::string strLine;
+    while (!strDump.eof() && strLine != HEADER_END)
+        getline(strDump, strLine); // Skip past header
+
+    std::string keyHex, valueHex;
+    while (!strDump.eof() && keyHex != DATA_END) {
+        getline(strDump, keyHex);
+        if (keyHex != DATA_END) {
+            if (strDump.eof())
+                break;
+            getline(strDump, valueHex);
+            if (valueHex == DATA_END) {
+                LogPrintf("Salvage: WARNING: Number of keys in data does not match number of values.\n");
+                break;
+            }
+            salvagedData.push_back(make_pair(ParseHex(keyHex), ParseHex(valueHex)));
+        }
+    }
+
+    if (keyHex != DATA_END) {
+        LogPrintf("Salvage: WARNING: Unexpected end of file while reading salvage output.\n");
+        fSuccess = false;
+    } else {
+        fSuccess = (result == 0);
+    }
+
+    if (salvagedData.empty())
+    {
+        LogPrintf("Salvage(aggressive) found no records in %s.\n", newFilename);
+        return false;
+    }
+    LogPrintf("Salvage(aggressive) found %u records\n", salvagedData.size());
+
+    std::unique_ptr<Db> pdbCopy = MakeUnique<Db>(env->dbenv.get(), 0);
+    int ret = pdbCopy->open(nullptr,               // Txn pointer
+                            filename.c_str(),   // Filename
+                            "main",             // Logical db name
+                            DB_BTREE,           // Database type
+                            DB_CREATE,          // Flags
+                            0);
+    if (ret > 0) {
+        LogPrintf("Cannot create database file %s\n", filename);
+        pdbCopy->close(0);
+        return false;
+    }
+
+    DbTxn* ptxn = env->TxnBegin();
+    CWallet dummyWallet(nullptr, WalletLocation(), WalletDatabase::CreateDummy());
+    for (KeyValPair& row : salvagedData)
+    {
+        /* Filter for only private key type KV pairs to be added to the salvaged wallet */
+        CDataStream ssKey(row.first, SER_DISK, CLIENT_VERSION);
+        CDataStream ssValue(row.second, SER_DISK, CLIENT_VERSION);
+        std::string strType, strErr;
+        bool fReadOK;
+        {
+            // Required in LoadKeyMetadata():
+            LOCK(dummyWallet.cs_wallet);
+            fReadOK = ReadKeyValue(&dummyWallet, ssKey, ssValue, strType, strErr);
+        }
+        if (!WalletBatch::IsKeyType(strType) && strType != DBKeys::HDCHAIN) {
+            continue;
+        }
+        if (!fReadOK)
+        {
+            LogPrintf("WARNING: WalletBatch::Recover skipping %s: %s\n", strType, strErr);
+            continue;
+        }
+        Dbt datKey(&row.first[0], row.first.size());
+        Dbt datValue(&row.second[0], row.second.size());
+        int ret2 = pdbCopy->put(ptxn, &datKey, &datValue, DB_NOOVERWRITE);
+        if (ret2 > 0)
+            fSuccess = false;
+    }
+    ptxn->commit(0);
+    pdbCopy->close(0);
+
+    return fSuccess;
```

Comment:
35a1f948 `RecoverDatabaseFile` contains a fair amount of code to set and return bool `fSuccess`, but the value returned doesn't appear to be used. Is this logic needed?

Reply:
The return value is being checked now.